libtenzir/builtins/formats/lines.cpp

```diff
@@ -106,44 +108,133 @@ class lines_parser final : public plugin_parser {
   parser_args args_;
 };
 
-// To get better diagnostics in `write`, i.e.:
-//    "'xsv' does not support heterogeneous outputs..."
-// -> "'lines' does not..."
-// Otherwise, we could just use the `ssv_plugin` directly
-class lines_printer final : public plugin_printer {
-public:
-  lines_printer() = default;
-
-  lines_printer(std::unique_ptr<plugin_printer> inner)
-    : inner_(std::move(inner)) {
+struct lines_printer_impl {
+  template <typename It>
+  auto print_values(It& out, const view<record>& x) const noexcept -> bool {
+    auto first = true;
+    for (const auto& [_, v] : x) {
+      if (!first) {
+        ++out = ' ';
+      } else {
+        first = false;
+      }
+      caf::visit(visitor{out}, v);
+    }
+    return true;
   }
 
+  template <class Iterator>
+  struct visitor {
+    visitor(Iterator& out) : out{out} {
+    }
+
+    auto operator()(caf::none_t) noexcept -> bool {
+      sequence_empty = false;
+      *out++ = '-'; // TODO: Should null values be mapped to empty string?
+      return true;
+    }
+
+    auto operator()(auto x) noexcept -> bool {
+      sequence_empty = false;
+      make_printer<decltype(x)> p;
+      return p.print(out, x);
+    }
+
+    auto operator()(view<pattern>) noexcept -> bool {
+      TENZIR_UNREACHABLE();
+    }
+
+    auto operator()(view<map>) noexcept -> bool {
+      TENZIR_UNREACHABLE();
+    }
+
+    auto operator()(view<std::string> x) noexcept -> bool {
+      sequence_empty = false;
+      out = std::copy(x.begin(), x.end(), out);
+      return true;
+    }
+
+    auto operator()(view<blob> x) noexcept -> bool {
+      return (*this)(detail::base64::encode(x));
+    }
+
+    auto operator()(const view<list>& x) noexcept -> bool {
+      sequence_empty = true;
+      for (const auto& v : x) {
+        if (!sequence_empty) {
+          ++out = ',';
```

Comment:
Should lists use a space as a separator?

Reply:
I think using `,` is as space is still (somewhat) linked to separate fields, rather than inside a field. Feel free to share if you think space is better.

Reply:
Yeah, I don't have a particularly strong opinion here. Since there can be no records inside lists and no nested lists after the call to `flatten` I think we can just choose what we want here. So let's stick with the comma.